src/controller/java/OTAProviderDelegateBridge.cpp

```diff
@@ -0,0 +1,409 @@
+                // session with another requestor when we get this query image request. We do not want to interrupt the
+                // ongoing transfer instead just respond to the second requestor with a busy status and a delayedActionTime
+                // in which the requestor can retry.
+                ExitNow(errorStatus = static_cast<uint8_t>(errorResponse.status));
+            }
+            LogErrorOnFailure(err);
+            commandObj->AddStatus(cachedCommandPath, StatusIB(err).mStatus);
+            // We need to reset state here to clean up any initialization we might have done including starting the BDX
+            // timeout timer while preparing for transfer if any failure occurs afterwards.
+            mBdxOTASender->ResetState();
+            ExitNow(errorStatus = static_cast<uint8_t>(OTAQueryStatus::kNotAvailable));
+        }
+        err = bdx::MakeURI(ourNodeId.GetNodeId(), jniFilePath.c_str() != nullptr ? jniFilePath.charSpan() : CharSpan(), uri);
+        if (CHIP_NO_ERROR != err)
+        {
+            LogErrorOnFailure(err);
+            commandObj->AddStatus(cachedCommandPath, StatusIB(err).mStatus);
+            mBdxOTASender->ResetState();
+            ExitNow(errorStatus = static_cast<uint8_t>(OTAQueryStatus::kNotAvailable));
+        }
+        response.imageURI.SetValue(uri);
+        commandObj->AddResponse(cachedCommandPath, response);
+    }
+    return;
+
+exit:
+    ChipLogError(Controller, "OTA Query Failure : %u, %" CHIP_ERROR_FORMAT, errorStatus, err.Format());
+    sendOTAQueryFailure(errorStatus);
+}
+
+void OTAProviderDelegateBridge::HandleApplyUpdateRequest(CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+                                                         const ApplyUpdateRequest::DecodableType & commandData)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturn(mOtaProviderDelegate.HasValidObjectRef(),
+                   commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::Failure));
+
+    NodeId nodeId = commandObj->GetSubjectDescriptor().subject;
+
+    ConcreteCommandPath cachedCommandPath(commandPath.mEndpointId, commandPath.mClusterId, commandPath.mCommandId);
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+
+    jobject otaProviderDelegate = mOtaProviderDelegate.ObjectRef();
+    chip::JniLocalReferenceManager manager(env);
+    jmethodID handleApplyUpdateRequestMethod;
+    CHIP_ERROR err = JniReferences::GetInstance().FindMethod(env, otaProviderDelegate, "handleApplyUpdateRequest",
+                                                             "(JJ)Lchip/devicecontroller/OTAProviderDelegate$ApplyUpdateResponse;",
+                                                             &handleApplyUpdateRequestMethod);
+    VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find handleApplyUpdateRequest method"));
```

Comment:
when error happens, should we commandObj->AddResponse with bad status?

Reply:
Modified.